pkg/open/open.go

```diff
@@ -0,0 +1,238 @@
+package open
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/glasskube/glasskube/api/v1alpha1"
+	"github.com/glasskube/glasskube/pkg/client"
+	"github.com/glasskube/glasskube/pkg/future"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+type opener struct {
+	pkgClient  *client.PackageV1Alpha1Client
+	ksClient   kubernetes.Interface
+	restClient rest.Interface
+	stopCh     []chan struct{}
+	readyCh    []chan struct{}
+	stopped    bool
+}
+
+func NewOpener() *opener {
+	return &opener{}
+}
+
+func (o *opener) Open(ctx context.Context, packageName string, entrypointName string) (*openResult, error) {
+	if err := o.initFromContext(ctx); err != nil {
+		return nil, err
+	}
+
+	var pkg v1alpha1.Package
+	if err := o.pkgClient.Packages().Get(ctx, packageName, &pkg); err != nil {
+		return nil, fmt.Errorf("could not get package %v: %w", packageName, err)
+	}
+
+	var packageInfo v1alpha1.PackageInfo
+	if err := o.pkgClient.PackageInfos().Get(ctx, pkg.Spec.PackageInfo.Name, &packageInfo); err != nil {
+		return nil, fmt.Errorf("could not get PackageInfo for package %v: %w", packageName, err)
+	}
+
+	entrypoints := packageInfo.Status.Manifest.Entrypoints
+	if len(entrypoints) < 1 {
+		return nil, fmt.Errorf("package %v has no entrypoint", packageName)
+	}
+
+	if entrypointName != "" {
+		exists := false
+		for _, entrypoint := range entrypoints {
+			if entrypoint.Name == entrypointName {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			return nil, fmt.Errorf("package %v has no entrypoint %v", packageName, entrypointName)
+		}
+	}
+
+	result := openResult{opener: o}
+	var futures []future.Future
+	for _, entrypoint := range entrypoints {
+		if entrypointName == "" || entrypoint.Name == entrypointName {
```

Comment:
cyclops requires the backend _and_ the frontend service to be get port-forwarded in other to work correctly. The backend actually get called via [`localhost:8080`](https://github.com/cyclops-ui/cyclops/blob/5637e47724be140481150767683239ef8abdd031/install/cyclops-install.yaml#L257)

We might need to forward multiple entrypoints, but only the first entry-point will get opened?  (e.g. `glasskube open cyclops ui ctrl`)

Reply:
Yes, this is true. That's why, by default, all entrypoints are forwarded and the first one is opened in the browser. This works for cyclops, once https://github.com/glasskube/packages/pull/13 gets merged.

I implemented it this way so we don't have to mess with the schema too much …